### ISSUE TITLE
Czech carriers

### ISSUE THREAD

**Issue by BronislavKlucka** [mmende/libphonenumber-geo-carrier#7]
Hello, 
your project has issues decoding Czech phone numbers

```
const main2 = async () => {
  const mobileNumber = parsePhoneNumberFromString('+420605582922')
  console.log(mobileNumber.formatInternational()); // +420 605 582 922
  const carrierEN = await carrier(mobileNumber) // T-Mobile
  console.log(carrierEN);
  
  console.log('-------------------------');
  
  const mobileNumber2 = parsePhoneNumberFromString('+420779990001')
  console.log(mobileNumber2.formatInternational()); // +420 779 990 001
  const carrierEN2 = await carrier(mobileNumber2) // null, should also be T-Mobile
  console.log(carrierEN2);
}

main2();
```

The issue seems to be in `getPrefix`, that returns 3 digits (605 or 779 in above cases), the problem is, that the bson is defined as

```
{
  '605': 'T-Mobile',
  '77999': 'T-Mobile',
  ... and bunch of others 
}
```
 the prefix checking should not be based on international phone number format
 
 Bronislav Klučka

**Comment by mmende** [comment]
Hi @BronislavKlucka 
Thanks for the hint. The first part of the international formatted number matched exactly with the prefix for all numbers I've tested so far. I couldn't find any infos about the semantics of the international formatting yet and just guessed that the first part after the country code was always the full prefix.
It might also simply be an issue with the formatting itself which is why the spacing ends up being wrong. I have to investigate a bit further to find out why the formatting differs from the other numbers in this case.

Do you have any suggestions on how to improve the prefix matching?

Kind regards

**Comment by BronislavKlucka** [comment]
Hi @mmende 
I can confirm, that this spacing / international formatting 
+420 779 990 001
for Czech numbers is correct, but prefixes are bit of a mess here.

My suggestion would be something like:
1/ create all versions of phone number (without country prefix) from full number to single digit by cutting digits from the end of number
779990001
77999000
7799900
779990
77999
7799
779
77
7

2/ look for key in BSON from the longest (most specific) version of the number to the shortest until one exists

there are other solutions, but this one is probably the fastest...


also your solution does not cache the loaded BSON file, if I need to find carrier for 100 Czech numbers, it will open and read the same file from filesystem 100 times... (just cache it based on BSON filepath in some object)

and if you cache it, it does not have to by async(promise)... first read will be sync (little stagger) but other will be from memory, you can also allow for precache all BSONs or maybe just some based on prefix/country at the start of consumers script (export some precache functions)...

**Comment by mmende** [comment]
Thanks for the suggestion. Narrowing down the number until finding the longest match seems to work fine. I adjusted the algorithm to do just that by using nationalNumber (which returns the national (significant) number without country calling code and whitespaces) and the number you posted now also returns the correct original carrier as expected. Before I publish the fix I however want to do some more testing. I'll close this issue once the new version is out.

Adding caching and an option to preload certain parts of the metadata sounds good. I'll see if I can integrate it in one of the upcoming releases.

Thanks for your contribution

**Comment by mmende** [comment]
New version is published (v1.2.5)... Let me know if everything works fine now
If the problem should still exist feel free to reopen the issue
